audio-echo/app/src/main/java/com/google/sample/echo/MainActivity.java

```diff
@@ -38,6 +46,8 @@ protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
 
+        //get run-time permission...
+        requestResourcePermission();
         status_view = (TextView)findViewById(R.id.statusView);
         queryNativeAudioParameters();
```

Comment:
will look at this with new code changes after studying the JF's recommendation. The key is at the start button: as along as we get permission before we execute "start button" actions, app will not crash; if no permission when user click "start button", no call should go into JNI code.